### ISSUE TITLE
mkl cpardiso solver when root process does not own any rows

### ISSUE THREAD

**Issue by thartland** [mfem/mfem#4634]
Context -- I aim to project a matrix coming from an FE discretizations to dofs defined on a surface and then use a direct solver on it (e.g., CPardiso). It happens that the root process may not own any dofs on said surface and so may not own any rows of the projected matrix. This seems to be an issue for CPardiso or perhaps in how mfem interfaces to it.

I have written a simple code that exposes this issue. Here solving Ax = b, where A is the identity matrix. I am able to use a GMRESSolver to obtain the solution when the root process does not own any rows but CPardiso fails. Specifically I get the error message:

```
Assertion failed: (error == 0) is false:
 --> CPardiso analyze input error
 ... in function: void mfem::CPardisoSolver::SetOperator(const mfem::Operator &)
 ... in file: linalg/cpardiso.cpp:154
```

I checked the value of `error` and it is `-1` which indicates "input inconsistent", see [here](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-fortran/2024-0/cluster-sparse-solver.html). The mumps solver does not have the same issue.

Any feedback would be appreciated. 

Here is the code which can be used to produce the issue with an mfem build in which MFEM_USE_MKL_CPARDISO=YES.

[CPardisoTest.txt](https://github.com/user-attachments/files/18186625/CPardisoTest.txt)

**Comment by adam-sim-dev** [comment]
I can reproduce this issue on Windows. The only bug I find is that, 

https://github.com/mfem/mfem/blob/701d3875c55f8179e795425fc6737d7e8f501b14/linalg/cpardiso.cpp#L130

should be:

```
if(m_loc==0)
{
  iparm[41] = first_row;
}
else
{
  iparm[41] = first_row + m_loc - 1;
}
```
Does MKL cluster_sparse_solver support the corner case where for one process it does not owns any rows? After reading the develop reference https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2023-0/cluster-sparse-solver.html, I did not find any useful information about it.

**Comment by psocratis** [comment]
@adam-sim-dev 

I did take a look at it a few days ago and what I noticed is that cpardiso has an issue if `iparm[41]` becomes negative which can only happen if the first N number of processors are empty. So similar to your suggestion I used something like 
````cpp
   if (first_row == 0 && m_loc == 0)
   {
      iparm[40] = 1;
      iparm[41] = 0;
   }
````
and this fixes the issue. I am not sure if this is the right fix though.

**Comment by adam-sim-dev** [comment]
> @adam-sim-dev
> 
> I did take a look at it a few days ago and what I noticed is that cpardiso has an issue if `iparm[41]` becomes negative which can only happen if the first N number of processors are empty. So similar to your suggestion I used something like
> 
> ```c++
>    if (first_row == 0 && m_loc == 0)
>    {
>       iparm[40] = 1;
>       iparm[41] = 0;
>    }
> ```
> 
> and this fixes the issue. I am not sure if this is the right fix though.

Yes, this works for this corner case. Do you agree it's a bug in MKL cluster_sparse_solver?  iparm[40] = 1 does not follow the developer reference.

**Comment by adam-sim-dev** [comment]
> > @adam-sim-dev
> > I did take a look at it a few days ago and what I noticed is that cpardiso has an issue if `iparm[41]` becomes negative which can only happen if the first N number of processors are empty. So similar to your suggestion I used something like
> > ```c++
> >    if (first_row == 0 && m_loc == 0)
> >    {
> >       iparm[40] = 1;
> >       iparm[41] = 0;
> >    }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and this fixes the issue. I am not sure if this is the right fix though.
> 
> Yes, this works for this corner case, I just tested it. Do you agree it's a bug in MKL cluster_sparse_solver? iparm[40] = 1 does not follow the developer reference.

**Comment by psocratis** [comment]
> > Yes, this works for this corner case, I just tested it. Do you agree it's a bug in MKL cluster_sparse_solver? iparm[40] = 1 does not follow the developer reference.

Yes, I think so.

**Comment by thartland** [comment]
Thanks for the input and the PR. I had discussed this with Socratis and the fix seems equivalent and so this seems to fix everything on the mfem side of things. Thanks again.
